### PR TITLE
New documentation for compile-time lists

### PR DESCRIPTION
**Pull request by Dicebot**
Intended to be merged after https://github.com/D-Programming-Language/phobos/pull/3128
and replace existing http://dlang.org/tuple

After that http://dlang.org/tuple can be rewritten to explain usage of `std.typecons.Tuple` in more
details and provide reference about historical naming reasons.

I am very sloppy with DDOC so any advice in improving formatting and style is very welcome

### CODE REVIEW

File: ctarguments.dd [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+ )
+
+$(P Consider this simple snippet:)
+
+---
+template Variadic(T...) { /* ... */ }
+---
+
+$(P `T` here is variadic $(LINK2 template.html#TemplateArgumentList, template argument list)
+    which is a core language feature. It is has own very special semantics
```
Comment (JinShil) [inline ctarguments.dd]: "is has"

File: ctarguments.dd [hunk 2]
```diff
@@ -0,0 +1,217 @@
+Ddoc
+
+$(D_S Compile-time (argument) lists,
+
+$(P Compile-time lists is a very important metaprogramming concept that comes naturally
+    from D support for $(LINK2 variadic-function-templates.html, variadic templates). Those
+    allow programmer to operate on types, symbols and expression allowing to define
```
Comment (JinShil) [inline ctarguments.dd]: "a programmer" or "programmers"

File: ctarguments.dd [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+$(P Consider this simple snippet:)
+
+---
+template Variadic(T...) { /* ... */ }
+---
+
+$(P `T` here is variadic $(LINK2 template.html#TemplateArgumentList, template argument list)
+    which is a core language feature. It is has own very special semantics
+    and from the programmer point of view is most similar to array of compile-time entities - types,
```
Comment (JinShil) [inline ctarguments.dd]: "an array"

File: ctarguments.dd [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    pragma(msg, T[1]);
+}
+
+alias Dummy = Variadic!(int, 42);
+// prints during compilation:
+// int
+// 42
+---
+
+$(P However, language itself does not provide any means to define such lists outside of
```
Comment (JinShil) [inline ctarguments.dd]: "language" -> "the language"

File: ctarguments.dd [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    template parameter declaration. Instead, there is a
+    $(LINK2 phobos/std_meta_arglist.html, simple utility) provided by D standard library:
+ )
+
+---
+alias Arguments(T...) = T;
+---
+
+$(P All it does is giving specific variadic argument list externally accessible name so
```
Comment (JinShil) [inline ctarguments.dd]: "giving" --> "give"
"a specific variadic argument list" or "specific variadic argument lists"
"an externally accessible name" or "externally accessible names"

File: ctarguments.dd [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+import std.meta.arglist;
+alias Numbers = Arguments!(1, 2, 3, 4);
+static assert (Numbers[1] == 2);
+alias SubNumbers = Numbers[1 .. $];
+static assert (SubNumbers[0] == 2);
+---
+
+$(H4 Assignment)
+
+$(P Works only if list element is a symbol that refers to a mutable variable)
```
Comment (JinShil) [inline ctarguments.dd]: "list element" --> "the list element"

File: ctarguments.dd [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    List[1] = 42;
+    assert (x == 42);
+    // List[0] = 42; // won't compile, can't assign to a constant
+}
+---
+
+$(H4 Loops)
+
+$(P D $(LINK2 dlang.org/statement.html#ForeachStatement, foreach statement) has special
+    semantics when iterating over compile-time list. It duplicates body of the loop
```
Comment (JinShil) [inline ctarguments.dd]: "body" --> "the body"

File: ctarguments.dd [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    for each of the list elements, with loop iteration "variable" becoming an alias
```
Comment (JinShil) [inline ctarguments.dd]: "loop iteration" --> "the loop iteration"

File: ctarguments.dd [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    for that compile-time list elements.
```
Comment (JinShil) [inline ctarguments.dd]: "the compile-time list elements" or "those compile-time list elements"

File: ctarguments.dd [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+   int
+   string
+   void()
+*/
+---
+
+$(H3 Auto-expansion)
+
+$(P One of less obvious properties of compile-time argument lists is that when used
```
Comment (JinShil) [inline ctarguments.dd]: "one of less" --> "one less"

File: ctarguments.dd [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+    as an argument to function or template those are automatically treated as list
```
Comment (JinShil) [inline ctarguments.dd]: "a function or template" or "functions or templates"
" a list of comma-separated..." or "lists of comma-separated..."

File: ctarguments.dd [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+template Print0(T...)
+{
+    pragma(msg, T[0]);
+}
+
+alias Dummy = Print0!(Arguments!(int, double));
+---
+
+$(P This will only print `int` during compilation because last line gets rewritten
```
Comment (JinShil) [inline ctarguments.dd]: "last line" --> "the last line"

File: ctarguments.dd [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+    as `alias Dummy = Print0!(int, double)`. If auto-expansion didn't happen,
+    `Arguments!(int, double)` would be printed instead. This is inherent part of
```
Comment (JinShil) [inline ctarguments.dd]: "inherent part" --> "an inherent part"

File: ctarguments.dd [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+    language semantics for variadic lists and thus also preserved by `Arguments`.
+ )
+
+$(H3 Homogenous lists)
+
+$(P `Arguments` that only consists from only type or expression (value) elements is
```
Comment (JinShil) [inline ctarguments.dd]: "consists" --> "consist"
the two "only"s sounds redundant.

File: ctarguments.dd [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+    commonly called "type list" or "expression list" accordingly. Concept of
```
Comment (JinShil) [inline ctarguments.dd]: type list  --> a "type list"
expression list --> an "expression list"
"Concept" --> "The concept"

File: ctarguments.dd [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+
+$(P It is possible to use homogenous type lists in declarations:)
+
+---
+import std.meta.arglist;
+alias Params = Arguments!(int, double, string);
+void foo(Params); // void foo(int, double, string);
+---
+
+$(P D supports special variable declaration syntax where type list acts as a type:)
```
Comment (JinShil) [inline ctarguments.dd]: "special variable declaration" --> "a special variable declaration"
"a type list" or "type lists"

File: ctarguments.dd [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+/* Compiler will rewrite such declaration to something like this:
+
+   int __var1;
+   double __var2;
+   string __var3;
+   alias variables = Arguments!(__var1, __var2, __var3);
+*/
+---
+
+$(P This is also what happens when you declare variadic template function:)
```
Comment (JinShil) [inline ctarguments.dd]: "a variadic template function" or "variadic template functions"

File: ctarguments.dd [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+
+---
+void foo(T...)(T args)
+{
+    // 'args' here is a compile-time list of symbols that
+    // refer to underlying compiler-generated arguments
```
Comment (JinShil) [inline ctarguments.dd]: "underlying compiler-generated arguments" --. "the underlying compiler-generated arguments"

File: ctarguments.dd [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+}
+---
+
+$(P It is possible to use expression lists with values of same type to declare array literals:)
```
Comment (JinShil) [inline ctarguments.dd]: "same type" --> "the same type"

**Comment by JinShil** [comment]
@CyberShadow  ctarguments.html does not appear in the documentation auto tester, so it is difficult to DDoc and formatting.

**Comment by andralex** [comment]
@JinShil okay to pull this now so you can submit edits as a separate PR?

**Comment by JinShil** [comment]
> @JinShil okay to pull this now so you can submit edits as a separate PR?

@andralex  Sure, I'd be happy to submit edits as a separate PR as long as it is reviewed and pulled in a timely manner.

**Comment by andralex** [comment]
done

**Pull request merged**